Subject: libdsm: use poll for POSIX NetBIOS sockets and abort pipes

Avoid indexing fixed-size fd_set objects with valid descriptors >= 1024.
Keep finite receive timeouts and cancellation wakeups across all NetBIOS waits.
Windows keeps its socket-handle based select implementation.

--- a/src/netbios_utils.h
+++ b/src/netbios_utils.h
@@ -68,4 +68,10 @@
 bool netbios_abort_ctx_is_aborted(struct netbios_abort_ctx *ctx);
 void netbios_abort_ctx_abort(struct netbios_abort_ctx *ctx);
 
+struct timeval;
+/* Returns 1 for socket readiness, 0 for timeout, or -1 for error/abort.
+ * A finite timeout is reduced by the elapsed wait, including spurious replies. */
+int netbios_wait(struct netbios_abort_ctx *ctx, int fd, bool writing,
+                 struct timeval *timeout);
+
 #endif
--- a/src/netbios_utils.c
+++ b/src/netbios_utils.c
@@ -35,6 +35,15 @@
 #include <string.h>
 #include <ctype.h>
 #include <stdlib.h>
+#include <errno.h>
+#include <limits.h>
+#include <time.h>
+#ifdef _WIN32
+#include <winsock2.h>
+#else
+#include <poll.h>
+#include <sys/time.h>
+#endif
 
 #include "netbios_utils.h"
 
@@ -180,16 +189,9 @@
 
 bool netbios_abort_ctx_is_aborted(struct netbios_abort_ctx *ctx)
 {
-    fd_set        read_fds;
-    int           res;
-    struct timeval timeout = {0, 0};
-
-    FD_ZERO(&read_fds);
-    FD_SET(ctx->pipe[0], &read_fds);
-
-    res = select(ctx->pipe[0] + 1, &read_fds, NULL, NULL, &timeout);
-
-    return (res < 0 || FD_ISSET(ctx->pipe[0], &read_fds));
+    struct pollfd fd = { .fd = ctx->pipe[0], .events = POLLIN };
+    int res = poll(&fd, 1, 0);
+    return res < 0 || fd.revents != 0;
 }
 
 void netbios_abort_ctx_abort(struct netbios_abort_ctx *ctx)
@@ -223,3 +225,89 @@
 }
 
 #endif
+
+int netbios_wait(struct netbios_abort_ctx *ctx, int fd, bool writing,
+                 struct timeval *timeout)
+{
+    if (fd < 0)
+    {
+        errno = EBADF;
+        return -1;
+    }
+#ifdef _WIN32
+    /* Winsock fd_set stores socket handles rather than indexing their values. */
+    fd_set read_fds, write_fds, error_fds;
+    FD_ZERO(&read_fds);
+    FD_ZERO(&write_fds);
+    FD_ZERO(&error_fds);
+    if (writing)
+        FD_SET(fd, &write_fds);
+    else
+        FD_SET(fd, &read_fds);
+    FD_SET(fd, &error_fds);
+    int res = select(fd + 1, &read_fds, &write_fds, &error_fds, timeout);
+    if (netbios_abort_ctx_is_aborted(ctx))
+    {
+        errno = EINTR;
+        return -1;
+    }
+    return res > 0 ? 1 : res;
+#else
+    struct pollfd fds[2] = {
+        { .fd = fd, .events = writing ? POLLOUT : POLLIN },
+        { .fd = -1, .events = POLLIN },
+    };
+#ifdef NS_ABORT_USE_PIPE
+    fds[1].fd = ctx->pipe[0];
+#else
+    if (netbios_abort_ctx_is_aborted(ctx))
+    {
+        errno = EINTR;
+        return -1;
+    }
+#endif
+    int milliseconds = -1;
+    int64_t remaining = 0;
+    struct timespec before;
+    if (timeout != NULL)
+    {
+        remaining = (int64_t)timeout->tv_sec * 1000000 + timeout->tv_usec;
+        int64_t rounded = (remaining + 999) / 1000;
+        milliseconds = rounded > INT_MAX ? INT_MAX : (int)rounded;
+        if (clock_gettime(CLOCK_MONOTONIC, &before) != 0)
+            return -1;
+    }
+    int res = poll(fds, 2, milliseconds);
+    int saved_errno = errno;
+    if (timeout != NULL)
+    {
+        struct timespec after;
+        if (clock_gettime(CLOCK_MONOTONIC, &after) != 0)
+            return -1;
+        int64_t elapsed = (int64_t)(after.tv_sec - before.tv_sec) * 1000000
+                         + (after.tv_nsec - before.tv_nsec) / 1000;
+        remaining = elapsed >= remaining ? 0 : remaining - elapsed;
+        timeout->tv_sec = remaining / 1000000;
+        timeout->tv_usec = remaining % 1000000;
+    }
+    errno = saved_errno;
+    if (res < 0)
+        return -1;
+#ifdef NS_ABORT_USE_PIPE
+    if (fds[1].revents != 0)
+#else
+    if (netbios_abort_ctx_is_aborted(ctx))
+#endif
+    {
+        errno = EINTR;
+        return -1;
+    }
+    if (fds[0].revents & POLLNVAL)
+    {
+        errno = EBADF;
+        return -1;
+    }
+    /* Let the subsequent send/recv/getsockopt report EOF or socket errors. */
+    return fds[0].revents != 0 ? 1 : 0;
+#endif
+}
--- a/src/netbios_ns.c
+++ b/src/netbios_ns.c
@@ -182,35 +182,14 @@
 
     while (true)
     {
-        fd_set read_fds, write_fds;
-        int nfds, ret;
-
-        FD_ZERO(&read_fds);
-        FD_ZERO(&write_fds);
-        FD_SET(ns->socket, &write_fds);
-#ifdef NS_ABORT_USE_PIPE
-        FD_SET(ns->abort_ctx.pipe[0], &read_fds);
-        nfds = (ns->socket > ns->abort_ctx.pipe[0] ? ns->socket : ns->abort_ctx.pipe[0]) + 1;
-#else
-        nfds = ns->socket + 1;
-#endif
-
-        ret = select(nfds, &read_fds, &write_fds, NULL, NULL);
+        int ret = netbios_wait(&ns->abort_ctx, ns->socket, true, NULL);
         if (ret < 0)
         {
-            BDSM_perror("netbios_ns_send_packet: select: ");
+            BDSM_perror("netbios_ns_send_packet: wait: ");
             return -1;
         }
 
-#ifdef NS_ABORT_USE_PIPE
-        if (FD_ISSET(ns->abort_ctx.pipe[0], &read_fds))
-            return -1;
-#else
-        if (netbios_abort_ctx_is_aborted(&ns->abort_ctx))
-            return -1;
-#endif
-
-        if (FD_ISSET(ns->socket, &write_fds))
+        if (ret > 0)
         {
             BDSM_dbg("Sending netbios packet to %s\n", inet_ntoa(addr.sin_addr));
             return sendto(ns->socket, (void *)q->packet,
@@ -463,45 +442,17 @@
     assert(ns != NULL);
 
     sock = ns->socket;
-#ifdef NS_ABORT_USE_PIPE
-    int abort_fd =  ns->abort_ctx.pipe[0];
-#else
-    int abort_fd = -1;
-#endif
 
     if (out_name_query)
         out_name_query->type = NAME_QUERY_TYPE_INVALID;
 
     while (true)
     {
-        fd_set read_fds, error_fds;
-        int res, nfds;
-
-        FD_ZERO(&read_fds);
-        FD_ZERO(&error_fds);
-        FD_SET(sock, &read_fds);
-#ifdef NS_ABORT_USE_PIPE
-        FD_SET(abort_fd, &read_fds);
-#endif
-        FD_SET(sock, &error_fds);
-        nfds = (sock > abort_fd ? sock : abort_fd) + 1;
-
-        res = select(nfds, &read_fds, 0, &error_fds, timeout);
-
+        int res = netbios_wait(&ns->abort_ctx, sock, false, timeout);
         if (res < 0)
             goto error;
-        if (FD_ISSET(sock, &error_fds))
-            goto error;
-
-#ifdef NS_ABORT_USE_PIPE
-        if (FD_ISSET(abort_fd, &read_fds))
-            return -1;
-#else
-        if (netbios_abort_ctx_is_aborted(&ns->abort_ctx))
-            return -1;
-#endif
-
-        else if (FD_ISSET(sock, &read_fds))
+
+        if (res > 0)
         {
             struct sockaddr_in addr;
             socklen_t addr_len = sizeof(struct sockaddr_in);
--- a/src/netbios_session.c
+++ b/src/netbios_session.c
@@ -96,32 +96,11 @@
         /* Wait for connection, cf. EINPROGRESS in man connect(2) */
         while (true)
         {
-            fd_set read_fds, write_fds;
-            int nfds, ret;
-
-            FD_ZERO(&read_fds);
-            FD_ZERO(&write_fds);
-            FD_SET(s->socket, &write_fds);
-#ifdef NS_ABORT_USE_PIPE
-            FD_SET(s->abort_ctx.pipe[0], &read_fds);
-            nfds = (s->socket > s->abort_ctx.pipe[0] ? s->socket : s->abort_ctx.pipe[0]) + 1;
-#else
-            nfds = s->socket + 1;
-#endif
-
-            ret = select(nfds, &read_fds, &write_fds, NULL, NULL);
+            int ret = netbios_wait(&s->abort_ctx, s->socket, true, NULL);
             if (ret < 0)
                 goto error;
 
-#ifdef NS_ABORT_USE_PIPE
-            if (FD_ISSET(s->abort_ctx.pipe[0], &read_fds))
-                return DSM_ERROR_GENERIC;
-#else
-            if (netbios_abort_ctx_is_aborted(&s->abort_ctx))
-                return DSM_ERROR_GENERIC;
-#endif
-
-            if (FD_ISSET(s->socket, &write_fds))
+            if (ret > 0)
             {
                 if (getsockopt(s->socket, SOL_SOCKET, SO_ERROR, &ret,
                                &(socklen_t){ sizeof (ret) }) || ret)
@@ -328,35 +307,14 @@
 
     while (true)
     {
-        fd_set read_fds, write_fds;
-        int nfds, ret;
-
-        FD_ZERO(&read_fds);
-        FD_ZERO(&write_fds);
-        FD_SET(s->socket, &write_fds);
-#ifdef NS_ABORT_USE_PIPE
-        FD_SET(s->abort_ctx.pipe[0], &read_fds);
-        nfds = (s->socket > s->abort_ctx.pipe[0] ? s->socket : s->abort_ctx.pipe[0]) + 1;
-#else
-        nfds = s->socket + 1;
-#endif
-
-        ret = select(nfds, &read_fds, &write_fds, NULL, NULL);
+        int ret = netbios_wait(&s->abort_ctx, s->socket, true, NULL);
         if (ret < 0)
         {
-            BDSM_perror("netbios_session_packet_send: select: ");
+            BDSM_perror("netbios_session_packet_send: wait: ");
             return 0;
         }
 
-#ifdef NS_ABORT_USE_PIPE
-        if (FD_ISSET(s->abort_ctx.pipe[0], &read_fds))
-            return 0;
-#else
-        if (netbios_abort_ctx_is_aborted(&s->abort_ctx))
-            return 0;
-#endif
-
-        if (FD_ISSET(s->socket, &write_fds))
+        if (ret > 0)
         {
             ssize_t sent = send(s->socket, (void *)s->packet, to_send, MSG_NOSIGNAL);
 
@@ -377,34 +335,14 @@
 {
     while (true)
     {
-        fd_set read_fds;
-        int nfds, ret;
-
-        FD_ZERO(&read_fds);
-        FD_SET(s->socket, &read_fds);
-#ifdef NS_ABORT_USE_PIPE
-        FD_SET(s->abort_ctx.pipe[0], &read_fds);
-        nfds = (s->socket > s->abort_ctx.pipe[0] ? s->socket : s->abort_ctx.pipe[0]) + 1;
-#else
-        nfds = s->socket + 1;
-#endif
-
-        ret = select(nfds, &read_fds, NULL, NULL, NULL);
+        int ret = netbios_wait(&s->abort_ctx, s->socket, false, NULL);
         if (ret < 0)
         {
-            BDSM_perror("netbios_session_recv: select: ");
+            BDSM_perror("netbios_session_recv: wait: ");
             return -1;
         }
 
-#ifdef NS_ABORT_USE_PIPE
-        if (FD_ISSET(s->abort_ctx.pipe[0], &read_fds))
-            return -1;
-#else
-        if (netbios_abort_ctx_is_aborted(&s->abort_ctx))
-            return -1;
-#endif
-
-        if (FD_ISSET(s->socket, &read_fds))
+        if (ret > 0)
         {
             ssize_t res = recv(s->socket, buf, len, 0);
 
